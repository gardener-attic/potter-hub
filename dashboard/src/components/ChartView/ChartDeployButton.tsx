import * as React from "react";
import { Redirect } from "react-router";
import { IChartVersion } from "../../shared/types";

interface IChartDeployButtonProps {
  version: IChartVersion;
  namespace: string;
}

interface IChartDeployButtonState {
  clicked: boolean;
}

class ChartDeployButton extends React.Component<IChartDeployButtonProps, IChartDeployButtonState> {
  public state: IChartDeployButtonState = {
    clicked: false,
  };

  public render() {
    const { version } = this.props;
    const { namespace } = this.props;
    const repoName = version.relationships.chart.data.repo.name;
    const chartName = version.relationships.chart.data.name;
    const versionStr = version.attributes.version;

    return (
      <div className="ChartDeployButton text-r">
        <button className="button button-primary button-accent" onClick={this.handleClick}>
          Deploy
        </button>
        {this.state.clicked && (
          <Redirect
            push={true}
            to={`/apps/ns/${namespace}/new/${repoName}/${chartName}/versions/${versionStr}`}
          />
        )}
      </div>
    );
  }

  private handleClick = () => {
    this.setState({ clicked: true });
  };
}

export default ChartDeployButton;
