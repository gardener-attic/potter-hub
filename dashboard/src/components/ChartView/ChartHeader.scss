.ChartHeader__button {
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
}
