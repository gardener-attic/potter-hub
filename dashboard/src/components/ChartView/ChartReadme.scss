.ChartReadme--error {
  flex-direction: column;
  height: 100%;
}

.ChartReadme p {
  margin: 0.625em 0;
}

.ChartReadme pre {
  margin: 0.625em 0;
  border-radius: 5px;
  overflow: auto;
}

.ChartReadme blockquote {
  margin: 0;
  text-align: left;
  font-size: 1em;
  border-radius: 5px;
  border: 1px solid #ddd;
  background-color: #f1f1f1;
  padding: 0.1em 1em;
  max-width: 100%;
}

.ChartReadme blockquote:before,
.ChartReadme blockquote:after {
  background-image: none;
}

.ChartReadme tr {
  border-bottom: 1px solid #eee;
}
