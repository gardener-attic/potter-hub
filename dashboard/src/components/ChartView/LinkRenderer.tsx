import * as React from "react";
import { HashLink as Link } from "react-router-hash-link";

const LinkRenderer: React.SFC<{}> = (props: any) => {
  if (props.href.startsWith("#")) {
    return <Link to={props.href}>{props.children}</Link>;
  }
  // If it's not a hash link it's an external link since it's rendering
  // the chart README. Because of that, render it as a normal anchor
  return <a href={props.href}>{props.children}</a>;
};

export default LinkRenderer;
