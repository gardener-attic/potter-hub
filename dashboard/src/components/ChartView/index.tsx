import ChartView from "./ChartView";

export default ChartView;
