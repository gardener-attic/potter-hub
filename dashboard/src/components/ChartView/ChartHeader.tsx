import * as React from "react";

import { IChartVersion } from "shared/types";
import ChartIcon from "../ChartIcon";
import ChartDeployButton from "./ChartDeployButton";
import "./ChartHeader.css";

interface IChartHeaderProps {
  id: string;
  icon?: string;
  repo: string;
  description: string;
  version: IChartVersion;
  namespace: string;
  showDeployButton: boolean;
}

class ChartHeader extends React.Component<IChartHeaderProps> {
  public render() {
    const { id, icon, description, version, namespace, showDeployButton } = this.props;
    return (
      <header>
        <div className="ChartView__heading margin-normal row">
          <div className="col-1 ChartHeader__icon">
            <ChartIcon icon={icon} />
          </div>
          <div className="col-9">
            <div className="title margin-l-small">
              <h1 className="margin-t-reset">{id}</h1>
              <h5 className="subtitle margin-b-reset">{description}</h5>
            </div>
          </div>
          { showDeployButton &&
            <div className="col-2 ChartHeader__button">
              <ChartDeployButton version={version} namespace={namespace} />
            </div>
          }
        </div>
        <hr />
      </header>
    );
  }
}

export default ChartHeader;
