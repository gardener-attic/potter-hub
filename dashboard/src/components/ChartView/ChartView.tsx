import * as React from "react";

import { IChartState, IChartVersion } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import ChartHeader from "./ChartHeader";
import ChartMaintainers from "./ChartMaintainers";
import ChartReadme from "./ChartReadme";
import ChartVersionsList from "./ChartVersionsList";
import "./ChartView.css";

interface IChartViewProps {
  chartID: string;
  fetchChartVersionsAndSelectVersion: (id: string, version?: string) => void;
  isFetching: boolean;
  selected: IChartState["selected"];
  selectChartVersion: (version: IChartVersion) => any;
  resetChartVersion: () => any;
  getChartReadme: (version: string) => any;
  namespace: string;
  version: string | undefined;
  targetClusterSecretNamespace: string;
  targetClusterSecretName: string;
}

class ChartView extends React.Component<IChartViewProps> {
  public componentDidMount() {
    const { chartID, fetchChartVersionsAndSelectVersion, version } = this.props;
    fetchChartVersionsAndSelectVersion(chartID, version);
  }

  public componentDidUpdate(prevProps: IChartViewProps) {
    const { selectChartVersion, version } = this.props;
    const { versions } = this.props.selected;
    if (prevProps.version !== version) {
      const cv = versions.find(v => v.attributes.version === version);
      if (cv) {
        selectChartVersion(cv);
      } else {
        throw new Error("could not find chart");
      }
    }
  }

  public componentWillUnmount() {
    this.props.resetChartVersion();
  }

  public render() {
    const { isFetching, getChartReadme, namespace, chartID } = this.props;
    const { version, readme, error, readmeError, versions } = this.props.selected;
    if (error) {
      return <ErrorSelector error={error} resource={`Chart ${chartID}`} />;
    }
    if (isFetching || !version) {
      return <LoadingWrapper />;
    }
    const chartAttrs = version.relationships.chart.data;
    return (
      <section className="ChartView padding-b-big">
        <ChartHeader
          id={`${chartAttrs.repo.name}/${chartAttrs.name}`}
          description={chartAttrs.description}
          icon={chartAttrs.icon}
          repo={chartAttrs.repo.name}
          version={version}
          namespace={namespace}
          showDeployButton={this.uiCalledForTargetCluster()}
        />
        <main>
          <div className="container container-fluid">
            <div className="row">
              <div className="col-9 ChartView__readme-container">
                <ChartReadme
                  getChartReadme={getChartReadme}
                  readme={readme}
                  hasError={!!readmeError}
                  version={version.attributes.version}
                />
              </div>
              <div className="col-3 ChartView__sidebar-container">
                <aside className="ChartViewSidebar bg-light margin-v-big padding-h-normal padding-b-normal">
                  <div className="ChartViewSidebar__section">
                    <h2>Chart Versions</h2>
                    <ChartVersionsList selected={version} versions={versions} />
                  </div>
                  {version.attributes.app_version && (
                    <div className="ChartViewSidebar__section">
                      <h2>App Version</h2>
                      <div>{version.attributes.app_version}</div>
                    </div>
                  )}
                  {chartAttrs.home && (
                    <div className="ChartViewSidebar__section">
                      <h2>Home</h2>
                      <div>
                        <ul className="remove-style padding-l-reset margin-b-reset">
                          <li>
                            <a href={chartAttrs.home} target="_blank" rel="noopener noreferrer">
                              {chartAttrs.home}
                            </a>
                          </li>
                        </ul>
                      </div>
                    </div>
                  )}
                  <div className="ChartViewSidebar__section">
                    <h2>Maintainers</h2>
                    <ChartMaintainers
                      maintainers={chartAttrs.maintainers}
                      githubIDAsNames={false}
                    />
                  </div>
                  {chartAttrs.sources?.length > 0 && (
                    <div className="ChartViewSidebar__section">
                      <h2>Related</h2>
                      <div className="ChartSources">
                        <ul className="remove-style padding-l-reset margin-b-reset">
                          {chartAttrs.sources.map((s, i) => (
                            <li key={i}>
                              <a href={s} target="_blank" rel="noopener noreferrer">
                                {s}
                              </a>
                            </li>
                          ))}
                        </ul>
                      </div>
                    </div>
                  )}
                </aside>
              </div>
            </div>
          </div>
        </main>
      </section>
    );
  }

  private uiCalledForTargetCluster = () => {
    return this.props.targetClusterSecretNamespace !== "" && this.props.targetClusterSecretName !== "";
  }

}

export default ChartView;
