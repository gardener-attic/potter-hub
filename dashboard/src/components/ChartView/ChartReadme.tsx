import * as React from "react";
import { FileText } from "react-feather";
import ReactMarkdown from "react-markdown";

import LoadingWrapper from "../LoadingWrapper";
import HeadingRenderer from "./HeadingRenderer";

import "./ChartReadme.css";
import LinkRenderer from "./LinkRenderer";

interface IChartReadmeProps {
  getChartReadme: (version: string) => void;
  hasError: boolean;
  readme?: string;
  version: string;
}

class ChartReadme extends React.Component<IChartReadmeProps> {
  public componentDidMount() {
    const { getChartReadme, version } = this.props;
    getChartReadme(version);
  }

  public componentDidUpdate(prevProps: IChartReadmeProps) {
    const { getChartReadme, version } = this.props;
    if (version !== prevProps.version) {
      getChartReadme(version);
    }
  }

  public render() {
    const { hasError, readme } = this.props;
    if (hasError) {
      return this.renderError();
    }
    return (
      <LoadingWrapper loaded={!!readme}>
        {readme && (
          <div className="ChartReadme">
            <ReactMarkdown
              source={readme}
              renderers={{
                heading: HeadingRenderer,
                link: LinkRenderer,
              }}
              skipHtml={true}
            />
          </div>
        )}
      </LoadingWrapper>
    );
  }

  public renderError() {
    return (
      <div className="ChartReadme ChartReadme--error flex align-center text-c">
        <FileText size={64} />
        <h4>No README found</h4>
      </div>
    );
  }
}

export default ChartReadme;
