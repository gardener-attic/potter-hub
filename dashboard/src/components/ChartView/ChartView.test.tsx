import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import itBehavesLike from "../../shared/specs";
import { IChartState, IChartVersion, NotFoundError } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import ChartHeader from "./ChartHeader";
import ChartMaintainers from "./ChartMaintainers";
import ChartReadme from "./ChartReadme";
import ChartVersionsList from "./ChartVersionsList";
import ChartView from "./ChartView";

const props: any = {
  chartID: "testrepo/test",
  fetchChartVersionsAndSelectVersion: jest.fn(),
  getChartReadme: jest.fn(),
  isFetching: false,
  namespace: "test",
  resetChartVersion: jest.fn(),
  selectChartVersion: jest.fn(),
  selected: {} as IChartState["selected"],
  version: undefined,
};

const testChart: IChartVersion["relationships"]["chart"] = {
  data: {
    repo: {
      name: "testrepo",
    },
  },
} as IChartVersion["relationships"]["chart"];

const testVersion = {
  attributes: {
    version: "1.2.3",
  },
  id: "1",
  relationships: { chart: testChart },
};

it("triggers the fetchChartVersionsAndSelectVersion when mounting", () => {
  const spy = jest.fn();
  shallow(<ChartView {...props} fetchChartVersionsAndSelectVersion={spy} />);
  expect(spy).toHaveBeenCalledWith("testrepo/test", undefined);
});

describe("when receiving new props", () => {
  it("finds and selects the chart version when version changes", () => {
    const versions = [{ attributes: { version: "1.2.3" } }] as IChartVersion[];
    const spy = jest.fn();
    const wrapper = shallow(
      <ChartView {...props} selectChartVersion={spy} selected={{ versions }} />,
    );
    wrapper.setProps({ version: "1.2.3" });
    expect(spy).toHaveBeenCalledWith(versions[0]);
  });

  it("does not trigger selectChartVersion if version does not change", () => {
    const spy = jest.fn();
    const wrapper = shallow(<ChartView {...props} selectChartVersion={spy} version="1.2.3" />);
    wrapper.setProps({ isFetching: true });
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it("throws an error if the chart version doesn't exist", () => {
    const versions = [{ attributes: { version: "1.2.3" } }] as IChartVersion[];
    const spy = jest.fn();
    const wrapper = shallow(
      <ChartView {...props} selectChartVersion={spy} selected={{ versions }} />,
    );
    expect(() => {
      wrapper.setProps({ version: "1.0.0" });
    }).toThrow("could not find chart");
  });
});

it("triggers resetChartVersion when unmounting", () => {
  const spy = jest.fn();
  const wrapper = shallow(<ChartView {...props} resetChartVersion={spy} />);
  wrapper.unmount();
  expect(spy).toHaveBeenCalled();
});

context("when fetching is false but no chart is available", () => {
  itBehavesLike("aLoadingComponent", {
    component: ChartView,
    props: {
      ...props,
      isFetching: false,
    },
  });
});

context("when fetching is true and chart is available", () => {
  itBehavesLike("aLoadingComponent", {
    component: ChartView,
    props: {
      ...props,
      isFetching: true,
      selected: { version: {} as IChartVersion },
    },
  });
});

describe("subcomponents", () => {
  const wrapper = shallow(<ChartView {...props} selected={{ version: testVersion }} />);

  for (const component of [ChartHeader, ChartReadme, ChartVersionsList, ChartMaintainers]) {
    it(`renders ${component.name}`, () => {
      expect(wrapper.find(component).exists()).toBe(true);
    });
  }
});

it("does not render the app version, home and sources sections if not set", () => {
  const wrapper = shallow(<ChartView {...props} selected={{ version: testVersion }} />);
  expect(wrapper.contains(<h2>App Version</h2>)).toBe(false);
  expect(wrapper.contains(<h2>Home</h2>)).toBe(false);
  expect(wrapper.contains(<h2>Related</h2>)).toBe(false);
});

it("renders the app version when set", () => {
  const v = testVersion as IChartVersion;
  v.attributes.app_version = "1.2.3-appversion";
  const wrapper = shallow(<ChartView {...props} selected={{ version: v }} />);
  expect(wrapper.contains(<h2>App Version</h2>)).toBe(true);
  expect(wrapper.contains(<div>1.2.3-appversion</div>)).toBe(true);
});

it("renders the home link when set", () => {
  const v = testVersion as IChartVersion;
  v.relationships.chart.data.home = "https://example.com";
  const wrapper = shallow(<ChartView {...props} selected={{ version: v }} />);
  expect(wrapper.contains(<h2>Home</h2>)).toBe(true);
  expect(
    wrapper.contains(
      <a href="https://example.com" target="_blank" rel="noopener noreferrer">
        {"https://example.com"}
      </a>,
    ),
  ).toBe(true);
});

describe("ChartMaintainers githubIDAsNames prop value", () => {
  const v = testVersion as IChartVersion;
  const tests: Array<{
    expected: boolean;
    name: string;
    repoURL: string;
  }> = [
    {
      expected: true,
      name: "stable Helm repo",
      repoURL: "https://kubernetes-charts.storage.googleapis.com",
    },
    {
      expected: true,
      name: "incubator Helm repo",
      repoURL: "https://kubernetes-charts-incubator.storage.googleapis.com",
    },
    { name: "random Helm repo", repoURL: "https://examplerepo.com", expected: false },
  ];

  for (const t of tests) {
    it(`for ${t.name}`, () => {
      v.relationships.chart.data.repo.url = t.repoURL;
      const wrapper = shallow(<ChartView {...props} selected={{ version: v }} />);
      const chartMaintainers = wrapper.find(ChartMaintainers);
      expect(chartMaintainers.props().githubIDAsNames).toBe(t.expected);
    });
  }
});

it("renders the sources links when set", () => {
  const v = testVersion as IChartVersion;
  v.relationships.chart.data.sources = ["https://example.com", "https://example2.com"];
  const wrapper = shallow(<ChartView {...props} selected={{ version: v }} />);
  expect(wrapper.contains(<h2>Related</h2>)).toBe(true);
  expect(
    wrapper.contains(
      <a href="https://example.com" target="_blank" rel="noopener noreferrer">
        {"https://example.com"}
      </a>,
    ),
  ).toBe(true);
  expect(
    wrapper.contains(
      <a href="https://example2.com" target="_blank" rel="noopener noreferrer">
        {"https://example2.com"}
      </a>,
    ),
  ).toBe(true);
});

describe("renders errors", () => {
  it("renders a not found error if it exists", () => {
    const wrapper = shallow(<ChartView {...props} selected={{ error: new NotFoundError() }} />);
    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(ErrorSelector).html()).toContain(`Chart ${props.chartID} not found`);
  });
  it("renders a generic error if it exists", () => {
    const wrapper = shallow(<ChartView {...props} selected={{ error: new Error() }} />);
    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(ErrorSelector).html()).toContain("Sorry! Something went wrong");
  });
});
