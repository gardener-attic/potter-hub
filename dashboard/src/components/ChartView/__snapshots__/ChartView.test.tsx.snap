// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when fetching is false but no chart is available loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`when fetching is true and chart is available loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;
